### PR TITLE
libfetchers: avoid api.github.com ratelimit if no github token is set

### PR DESCRIPTION
**Pull request by Mic92**
If we don't have any github token, we won't be able to fetch private repos, but we are also more likely to run into API limits since we don't have a token. To mitigate this only ever use the github api if we actually have a token.

This is an alternative to https://github.com/NixOS/nix/pull/6997

closes https://github.com/NixOS/nix/issues/5409

### CODE REVIEW

File: src/libfetchers/github.cc [hunk 1]
```diff
@@ -265,14 +265,18 @@ struct GitHubInputScheme : GitArchiveInputScheme
...
         auto host = maybeGetStrAttr(input.attrs, "host").value_or("github.com");
-        auto url = fmt(
-            host == "github.com"
-            ? "https://api.%s/repos/%s/%s/tarball/%s"
-            : "https://%s/api/v3/repos/%s/%s/tarball/%s",
-            host, getStrAttr(input.attrs, "owner"), getStrAttr(input.attrs, "repo"),
+        Headers headers = makeHeadersWithAuthTokens(host);
+        const char* urlFmt;
+        if (host == "github.com")
+            // If we have no auth headers than we default to the public archive
```
Comment (knedlsepp) [inline src/libfetchers/github.cc:271]: than -> then

File: src/libfetchers/github.cc [hunk 2]
```diff
@@ -265,14 +265,18 @@ struct GitHubInputScheme : GitArchiveInputScheme
         // FIXME: use regular /archive URLs instead? api.github.com
```
Comment (knedlsepp) [inline src/libfetchers/github.cc:265]: FIXME can be dropped now

File: src/libfetchers/github.cc [hunk 3]
```diff
@@ -265,14 +265,18 @@ struct GitHubInputScheme : GitArchiveInputScheme
...
         // might have stricter rate limits.
         auto host = maybeGetStrAttr(input.attrs, "host").value_or("github.com");
-        auto url = fmt(
-            host == "github.com"
-            ? "https://api.%s/repos/%s/%s/tarball/%s"
-            : "https://%s/api/v3/repos/%s/%s/tarball/%s",
-            host, getStrAttr(input.attrs, "owner"), getStrAttr(input.attrs, "repo"),
+        Headers headers = makeHeadersWithAuthTokens(host);
+        const char* urlFmt;
```
Comment (knedlsepp) [inline src/libfetchers/github.cc:269]: Any reason to not opt for a std::string here?

Comment (Mic92) [reply src/libfetchers/github.cc:269]: afaik, this would create an unnecessary string allocation. Also I cannot use const in this case.

Comment (edolstra) [reply src/libfetchers/github.cc:267]: I think `std::string_view` should work.

Comment (edolstra) [reply src/libfetchers/github.cc:267]: You can probably also write this as
```
auto urlFmt =
    host == "github.com"
    ? headers.empty() 
       ? "https://%s/%s/%s/archive/%s.tar.gz" 
       : "https://api.%s/repos/%s/%s/tarball/%s"
    : "https://%s/api/v3/repos/%s/%s/tarball/%s"
```

Comment (Mic92) [reply src/libfetchers/github.cc:267]: > You can probably also write this as
> 
> ```
> auto urlFmt =
>     host == "github.com"
>     ? headers.empty() 
>        ? "https://%s/%s/%s/archive/%s.tar.gz" 
>        : "https://api.%s/repos/%s/%s/tarball/%s"
>     : "https://%s/api/v3/repos/%s/%s/tarball/%s"
> ```

I find this quite hard to read to be honest.

Comment (Mic92) [reply src/libfetchers/github.cc:267]: https://github.com/NixOS/nix/pull/7039

Comment (Mic92) [reply src/libfetchers/github.cc:267]: Closed this pr and move ahead in https://github.com/NixOS/nix/pull/7039

**Pull request closed without merging**